main.go

```diff
@@ -1,204 +1,105 @@
-	<-done
-	setupLog.Info("Server stopped")
-
-	// Shutdown http server
-	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-	defer func() {
-		cancel()
-	}()
-	if err := srv.Shutdown(ctx); err != nil {
-		setupLog.Error(err, "Server shutdown failed")
-		os.Exit(1)
+		tokenWriter := token.NewTokenWriter(client)
+		go tokenWriter.Start(ctx.Done())
 	}
 
-	setupLog.Info("Server exited properly")
-}
-
-func readinessHandler(log logr.Logger) func(http.ResponseWriter, *http.Request) {
-	return func(w http.ResponseWriter, r *http.Request) {
-		w.WriteHeader(http.StatusOK)
-		w.Header().Set("Content-Type", "application/json")
-		if _, err := w.Write([]byte("{\"status\": \"ok\"}")); err != nil {
-			log.Error(err, "Could not write response data")
-		}
-	}
-}
+	setupLog.Info("Starting server")
+	azdoServer := server.NewAzdoServer(logger, strconv.Itoa(port), authz)
+	azdoServer.ListenAndServe(ctx.Done())
 
-func livenessHandler(log logr.Logger) func(http.ResponseWriter, *http.Request) {
-	return func(w http.ResponseWriter, r *http.Request) {
-		w.WriteHeader(http.StatusOK)
-		w.Header().Set("Content-Type", "application/json")
-		if _, err := w.Write([]byte("{\"status\": \"ok\"}")); err != nil {
-			log.Error(err, "Could not write response data")
-		}
-	}
+	setupLog.Info("Azure DevOps Proxy stopped")
 }
 
-func proxyHandler(p *httputil.ReverseProxy, log logr.Logger, a *auth.Authorization, domain string, pat string) func(http.ResponseWriter, *http.Request) {
-	return func(w http.ResponseWriter, r *http.Request) {
-		// If basic auth is missing return error to force client to retry
-		username, password, ok := r.BasicAuth()
-		if !ok {
-			w.Header().Set("WWW-Authenticate", "Basic realm=\"Restricted\"")
-			http.Error(w, "Missing basic authentication", http.StatusUnauthorized)
-			return
-		}
-
-		t := username
-		if len(password) > 0 {
-			t = password
+func getKubernetesClient(kubeconfigPath string) (*kubernetes.Clientset, error) {
+	if kubeconfigPath != "" {
+		cfg, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+		if err != nil {
+			return nil, err
 		}
-
-		// Check basic auth with local auth configuration
-		err := auth.IsPermitted(a, r.URL.EscapedPath(), t)
+		clientset, err := kubernetes.NewForConfig(cfg)
 		if err != nil {
-			log.Error(err, "Received unauthorized request")
-			http.Error(w, "User not permitted", http.StatusForbidden)
-			return
+			return nil, err
 		}
-
-		// Forward request to destination server
-		log.Info("Authenticated request", "path", r.URL.Path)
-		r.Host = domain
-		r.Header.Del("Authorization")
-		patB64 := base64.StdEncoding.EncodeToString([]byte("pat:" + pat))
-		r.Header.Add("Authorization", "Basic "+patB64)
-
-		p.ServeHTTP(w, r)
+		return clientset, nil
 	}
-}
 
-func validatePat(pat string, org string) error {
-	url := fmt.Sprintf("https://dev.azure.com/%v/_apis/connectionData", org)
-	req, err := http.NewRequest("GET", url, nil)
+	config, err := rest.InClusterConfig()
```

Comment:
To make it easier to use localy should we also support default kubeconfig file using k8s.io/client-go/util/homedir or your KUBECONFIG file?

Reply:
It's enough with the config path